src/modules/commander/state_machine_helper.cpp

```diff
@@ -912,6 +845,153 @@ void reset_link_loss_globals(actuator_armed_s *armed, const bool old_failsafe, c
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
+			return;
+		}
+
+	}
+
+	// If none of the above worked, try to mitigate
+	if (status_flags.condition_local_altitude_valid) {
+		if (status->vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_DESCEND;
+
+		} else {
+			// TODO: FW position controller doesn't run without condition_global_position_valid
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDGPSFAIL;
+		}
+
+	} else {
+		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
+	}
+}
+
+void set_offboard_loss_rc_nav_state(vehicle_status_s *status, actuator_armed_s *armed,
+				    const vehicle_status_flags_s &status_flags,
+				    const offboard_loss_rc_actions_t offboard_loss_rc_act)
+{
+	switch (offboard_loss_rc_act) {
+	case offboard_loss_rc_actions_t::DISABLED:
+		// If offboard loss failsafe is disabled then no action must be taken.
+		return;
+
+	case offboard_loss_rc_actions_t::TERMINATE:
+		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
+		armed->force_failsafe = true;
+		return;
+
+	case offboard_loss_rc_actions_t::LOCKDOWN:
+		armed->lockdown = true;
+		return;
+
+	case offboard_loss_rc_actions_t::MANUAL_POSITION:
+		if (status_flags.condition_global_position_valid) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_POSCTL;
+			return;
+		}
+
+	// FALLTHROUGH
+	case offboard_loss_rc_actions_t::MANUAL_ALTITUDE:
+		if (status_flags.condition_local_altitude_valid) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_ALTCTL;
+			return;
+		}
+
+	// FALLTHROUGH
+	case offboard_loss_rc_actions_t::MANUAL_ATTITUDE:
+		status->nav_state = vehicle_status_s::NAVIGATION_STATE_MANUAL;
+		return;
+
+	case offboard_loss_rc_actions_t::AUTO_RTL:
+		if (status_flags.condition_global_position_valid && status_flags.condition_home_position_valid) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_RTL;
+			return;
+		}
+
+	// FALLTHROUGH
+	case offboard_loss_rc_actions_t::AUTO_LOITER:
+		if (status_flags.condition_global_position_valid) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
+			return;
+		}
+
+	// FALLTHROUGH
+	case offboard_loss_rc_actions_t::AUTO_LAND:
+		if (status_flags.condition_global_position_valid) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
+			return;
+		}
+
+	}
+
+	// If none of the above worked, try to mitigate
+	if (status_flags.condition_local_altitude_valid) {
+		//TODO: Add case for rover
+		if (status->vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_DESCEND;
+
+		} else {
+			// TODO: FW position controller doesn't run without condition_global_position_valid
+			status->nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDGPSFAIL;
+		}
+
+	} else {
+		status->nav_state = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
```

Comment:
Does it need the `force_failsafe` here?

Reply:
No idea! I wanted to minimize the logic changes in this PR.

Reply:
I see, we can leave it as is and test and change it later.